### PR TITLE
osd: set min_version to newest version in maybe_force_recovery

### PR DESCRIPTION
**Pull request by XinzeChi**

### CODE REVIEW

**Comment by XinzeChi** [comment]
@tchaikov ping

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -728,7 +728,7 @@ void PrimaryLogPG::maybe_force_recovery()
   // find the oldest missing object
   version_t min_version = pg_log.get_log().head.version;
   hobject_t soid;
-  if (!pg_log.get_missing().get_items().empty()) {
+  if (!pg_log.get_missing().get_rmissing().empty()) {
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: > rmissing != missing in backfill

why?

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -726,7 +726,7 @@ void PrimaryLogPG::maybe_force_recovery()
     return;
 
   // find the oldest missing object
-  version_t min_version = 0;
+  version_t min_version = pg_log.get_log().head.version;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: could you elaborate a little bit the rationale of this change in the commit message?

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (XinzeChi) [reply src/osd/PrimaryLogPG.cc]: in prep_backfill_object_push func, we know the needed version of all backfilling object is zero

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (XinzeChi) [reply src/osd/PrimaryLogPG.cc]: if min_version is 0, if (min_version > xxx) always return false

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: how is this connected to the conclusion of `missing.empty() != rmissing.empty()`?

Comment (XinzeChi) [reply src/osd/PrimaryLogPG.cc]: because we use rmissing iterator instead of missing iterator, rmissing.empty() looks more correctly

Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: so, i think it's more of a cleanup? am i right?

Comment (XinzeChi) [reply src/osd/PrimaryLogPG.cc]: yes

Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: could you update the commit message accordingly?

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: the question is why `pg_log.get_log().head.version`  is correct, but `0` is not.

Comment (XinzeChi) [reply src/osd/PrimaryLogPG.cc]: what we want to get is the minimum of the objects' versions. so we should initialize to as head version and compare it with other version

Comment (tchaikov) [reply src/osd/PrimaryLogPG.cc]: oic, could you amend the commit message with the explanation above?

**Comment by xiexingguo** [comment]
instead of resetting min_version at the very beginning (and change these codes significantly), I think you can:

```diff
--- a/src/osd/PrimaryLogPG.cc
+++ b/src/osd/PrimaryLogPG.cc
@@ -723,6 +723,7 @@ void PrimaryLogPG::maybe_force_recovery()
     pg_shard_t peer = *it;
     if (peer_missing.count(peer) &&
        !peer_missing[peer].get_items().empty() &&
+        min_version &&
        min_version > peer_missing[peer].get_rmissing().begin()->first) {
       min_version = peer_missing[peer].get_rmissing().begin()->first;
       soid = peer_missing[peer].get_rmissing().begin()->second;
```

**Comment by tchaikov** [comment]
@xiexingguo why make `0` a special case? in the sense of correctness, i think to use the maximum possible max eversion as the initial `min_version` does make sense.

**Comment by xiexingguo** [comment]
>  use the maximum possible max eversion as the initial min_version does make sense.

Fine with me, kefu:-)

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -739,10 +739,14 @@ void PrimaryLogPG::maybe_force_recovery()
     if (*it == get_primary()) continue;
     pg_shard_t peer = *it;
     if (peer_missing.count(peer) &&
-	!peer_missing[peer].get_items().empty() &&
-	min_version > peer_missing[peer].get_rmissing().begin()->first) {
-      min_version = peer_missing[peer].get_rmissing().begin()->first;
-      soid = peer_missing[peer].get_rmissing().begin()->second;
+	!peer_missing[peer].get_items().empty()) {
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: might want to drop this change or split it into another separate "refactor only" commit.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-10-01_17:38:10-rados-wip-kefu-testing-2017-10-01-2202-distro-basic-mira/

**Pull request merged**